### PR TITLE
Under `-Xsource:3`, warn that inherited members no longer take precedence over outer definitions in Scala 3

### LINKED ISSUE #11921: Type error when outer scope name is shadowed by inheritance
## reproduction steps

```scala
Welcome to Scala 2.13.1 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_181).
Type in expressions for evaluation. Or try :help.

scala> def lazyMap[T, U](coll: Iterable[T], f: T => U) = 
     |   new Iterable[U] {
     |     def iterator = coll.iterator map f 
     |   }
```

## problem

```scala
scala> def lazyMap[T, U](coll: Iterable[T], f: T => U) = 
     |   new Iterable[U] {
     |     def iterator = coll.iterator map f
     |   }
           def iterator = coll.iterator map f
                                            ^
On line 3: error: type mismatch;
        found   : T => U
        required: U => U
```

## expectation
This code is from book *Programming in Scala 3rd Edition*.  Section `24.14 Views`

The type of coll should be Iterable[T] but complier judge it as Iterable[U] because the `coll` is as the same name the function `coll` defined in trait `Iterable`

In `src/library/scala/collection/Iterable.scala`

```scala
trait Iterable[+A] extends IterableOnce[A]
  with IterableOps[A, Iterable, Iterable[A]]
  with IterableFactoryDefaults[A, Iterable] {

  // The collection itself
  final def toIterable: this.type = this

  final protected def coll: this.type = this
....
```
I'm new to Scala, I'm not sure here is design as a feature or not.

One workaroud is rename the parameter
```scala
scala> def lazyMap[T, U](coll: Iterable[T], f: T => U) = {
     |   val _coll = coll
     |   new Iterable[U] {
     |     def iterator = _coll.iterator.map(f)
     |   }
     | }
lazyMap: [T, U](coll: Iterable[T], f: T => U)Iterable[U]
```

Consider the inner variable socpe will cover the outer ones, should add some hint here? Because a trait may have a lot of the function or variable, check variable inside defined trait one by one may not be so easy.

How about adding some hints like this?
```scala
scala> def lazyMap[T, U](coll: Iterable[T], f: T => U) = 
     |   new Iterable[U] {
     |     def iterator = coll.iterator map f
     |   }
           
           def lazyMap[T, U](coll: Iterable[T], f: T => U) = 
                                ^
On line 1: warning: variable defined in inner scope

     "coll" has defined in triat "Iterable", may be you need to rename it.

           def iterator = coll.iterator map f
                                            ^
On line 3: error: type mismatch;
        found   : T => U
        required: U => U
```

### PR DESCRIPTION
**Pull request by lrytz**
Continuation of @som-snytt's work at https://github.com/scala/scala/pull/10177.

Backport from Scala 3: an inherited name no longer takes precedence over a definition in an outer scope. The following reference `m` is ambiguous:

```scala
class C { def m = ??? }
def f(m: Int) = new C {
  def g = m   // use this.m
}
```

It's an error under `-Xsource:3`, a warning otherwise.

Fixes scala/bug#11921.

Don't issue an ambiguity error if
  - a symbol is a type alias and the two types are equivalent (`=:=`)
  - both symbols are stable and their singleton types are equivalent (`=:=`)

Example:

```scala
class C
class A { val c: B.c.type = B.c }
object B {
  val c = new C
  val a = new A {
    def m = c    // not ambiguous, `this.c` vs `B.c`
  }
}
```

Type example:

```scala
class C[T] { type TT = T }
object O {
  type TT = String
  class D extends C[TT] {
    def n(x: TT) = x  // not ambiguous, `this.TT` vs `O.TT`
  }
}
```

### CODE REVIEW

**Comment by som-snytt** [comment]
I added the emoji for Artemis rocket headed for the moon!

**Comment by lrytz** [comment]
Scala 3 variant: https://github.com/lampepfl/dotty/pull/16401

**Comment by som-snytt** [comment]
There must be a word for when you are able to contribute versions of a fix on both 2 and 3, coincidentally using the prefix "ambi-".

I propose "ambiversal".

**Comment by lrytz** [comment]
The most common pattern leading to spurious errors is this:

```scala
class C(val name: String)
object Test {
  def m(name: String) = new C(name) {
    println(name)
  }
}
```

I spent some time coming up with a heuristic to see through the most common such pattern [EDIT: but the heuristic was entirely discarded ultimately, so the example above warns].

The difficulty is, `new C(name) { ... }` is pulled apart by the type checker into `new $anon { def <init>() = { super.<init>(name); ... } ... }`, we don't get a typed AST that represents `new C(name)`. But at least the `Ident(name)` constructor argument that we see in the `template.parents` has a symbol assigned (I assume by chance, that's a side-effect of `typedIdent`).

I pushed this for now to give it a try. If we keep it, we should try to do something similar for Scala 3.

The heuristic currently breaks here:

```scala
abstract class A(val name: String)
class C(name: String) extends A(name)
object Test {
  def m(name: String) = new C(name) {
    println(name)
  }
}
```

**Comment by lrytz** [comment]
I think the last experiment is probably not worth it. The implementation is rather ad-hoc, and it would have to be replicated to more compilers (Scala 3, IntelliJ).

Instead I changed it to be a warning (WIP) and made it even more verbose.

**Review by som-snytt — APPROVED** [review]
Very nice!

**Comment by som-snytt** [comment]
This was on my to-do list yesterday, right after "override message" and "dentist" (for child). There was a cavity. 😢 

The code inherited from the other PR (`foundCompetingSymbol`) is not as clear as one would like.

The amendment to `nextDefinition` is very nice and doesn't introduce extra tests until required.

Triggering the warning in typer is especially clever yet simple. Just warning is probably sufficient, since `-Werror` is well advertised.

**Comment by lrytz** [comment]
Thanks, I should also learn writing supportive review messages.

**Comment by som-snytt** [comment]
For fun, I verified the warning for the linked PR.
```
[warn] .../scala/src/library/scala/collection/immutable/BitSet.scala:251:19: reference to empty is ambiguous;
[warn] it is both defined in the enclosing object BitSet and available in the enclosing class BitSetN as method empty (defined in class BitSet)
[warn] Since 3, symbols inherited from a superclass no longer shadow symbols defined in an outer scope.
[warn] To continue using the symbol from the superclass, write `this.empty`.
[warn] Or use `-Wconf:msg=legacy-binding:s` to silence this warning.
[warn]                   empty
[warn]                   ^
[warn] .../scala/src/library/scala/collection/immutable/BitSet.scala:330:15: reference to empty is ambiguous;
[warn] it is both defined in the enclosing object BitSet and available in the enclosing class BitSetN as method empty (defined in class BitSet)
[warn] Since 3, symbols inherited from a superclass no longer shadow symbols defined in an outer scope.
[warn] To continue using the symbol from the superclass, write `this.empty`.
[warn] Or use `-Wconf:msg=legacy-binding:s` to silence this warning.
[warn]               empty
[warn]               ^
```
The build has other warnings under JDK 19, so if you're inclined to ignore warnings, you might miss it. I also wondered if "Since 3" reads normally. "Since 3 P.M."

I thought I'd contributed something to mitigate the JDK19 warnings.

**Comment by lrytz** [comment]
`Since 3` is definitely a typo, will fix it.

**Comment by SethTisue** [comment]
@lrytz this caused https://scala-ci.typesafe.com/job/scala-2.13.x-jdk11-integrate-community-build/4184/artifact/logs/kafka-build.log — the source file is https://github.com/ennru/kafka/blob/eda4a2904a863b04f6ea511a4cf1e2bd4806ad35/core/src/test/scala/unit/kafka/server/ReplicaManagerTest.scala — is it a regression, or does it seem expected to you?

**Comment by som-snytt** [comment]
Seems unexpected. The `replicaManager` is a member (ctor arg) of the enclosing class and the instance is getting assigned to the conflicting definition. That is only a bit weird.

**Comment by som-snytt** [comment]
Simple repro (or actually it looks slightly different) (probably won't be able to pursue it at the moment)
```
trait T
abstract class C(val r: T)

class D {
  val r = new C(null) {
    def f = r
  }
}
```
compiles previously but
```
pr10220.scala:6: error: recursive method f needs result type
  val r = new C(null) {
          ^
pr10220.scala:7: warning: reference to r is ambiguous;
it is both defined in the enclosing class D and available in the enclosing anonymous class as value r (defined in class C)
Since 3, symbols inherited from a superclass no longer shadow symbols defined in an outer scope.
To continue using the symbol from the superclass, write `this.r`.
Or use `-Wconf:msg=legacy-binding:s` to silence this warning.
    def f = r
            ^
1 warning
1 error
```
Maybe Lukas knows how to do the alias probe less forcefully. Almost said alien probe.

**Comment by lrytz** [comment]
Thank you for the minimization @som-snytt!

> Maybe Lukas knows how to do the alias probe less forcefully

Hmm, didn't find a way so far... But it's probably a regression we can live with?

This fails to compile before my change (recursive method f needs result type):

```scala
trait T
abstract class C(val r: T)
class D {
  val r = new C(null) {
    def f = D.this.r
  }
}
```

**Comment by som-snytt** [comment]
Regressions are pretty easy to live with. They don't snore and always clean the sink when they are done in the morning.

I argued for erroring by default, so I would welcome more regressions. [Autocomplete suggests "terrorizing".]

If there is a place for `-Ylegacy-whatever`, it would be just to exclude this corner case. Possibly if a codebase contains this idiom, possibly only in tests, there could be a bunch of them. But extra options incur as much user wrath as regressions, perhaps more because of decision fatigue.

File: spec/02-identifiers-names-and-scopes.md [hunk 1]
```diff
@@ -17,14 +17,16 @@ which are collectively called _bindings_.
 Bindings of each kind are assigned a precedence which determines
 whether one binding can shadow another:
 
-1. Definitions and declarations that are local, inherited, or made
-   available by a package clause and also defined in the same compilation unit
-   as the reference to them, have the highest precedence.
+1. Definitions and declarations in lexical scope that are not [top-level](09-top-level-definitions.html)
```
Comment (lrytz) [inline spec/02-identifiers-names-and-scopes.md:20]: @som-snytt I assume `top-level` is excluded here because that's what's implemented in Scala 3 (https://github.com/lampepfl/dotty/pull/8622). In Scala 3, the following compiles and `new X` resolves to `C.this.X`

```scala
package p {
  trait T { class X }
  class C extends T {
    def test = new X
  }
  class X
}
```

Changing `package p` to `object p` makes `new X` ambiguous. Do you see a good reason for that? cc @odersky

Comment (som-snytt) [reply spec/02-identifiers-names-and-scopes.md:20]: If `T` has a member `def name: String = "John Doe"`, then just where it can be inherited, introducing a different definition of `name` might be in conflict.

Packagings are not classes and don't suffer from extension, so I should be able to write top-level `def name = "Gloria Mundi"` without concern about confusing anyone.

But this is ambiguous:
```
    trait T {
      class X
      def name: String = "John Doe"
    }
    class C extends T {
      def test = new X
      def result = name
    }
    class X
    //def name = "Gloria Mundi"
    val name: T = new T {
      def result = name
    }
```
I expected it to prefer the inherited `name` and not the enclosing definition which is a top-level def.
```
15 |      def result = name
   |                   ^^^^
   |                   Reference to name is ambiguous,
   |                   it is both defined in package object t10220$package
   |                   and inherited subsequently in anonymous class Object with p.T {...}
```
I don't think they actually got rid of package objects...

Comment (lrytz) [reply spec/02-identifiers-names-and-scopes.md:20]: So you expect an inherited symbol takes precedence over an outer definition in the same source file, if that outer definition is package-owned? Can you say why? I don't yet see what difference it makes if the outer symbol is top-level or not.

> I should be able to write top-level `def name = "Gloria Mundi"` without concern about confusing anyone

What is the difference in that regard between `package p { def name = "..." }` and `object p { def name = "..."}`?

Comment (lrytz) [reply spec/02-identifiers-names-and-scopes.md:20]: For the record, I'm on this topic because I'm testing https://github.com/scala/scala/pull/10297, and I see warnings that didn't show up in 2.13. These new warnings seem helpful to me.

Comment (som-snytt) [reply spec/02-identifiers-names-and-scopes.md:20]: I guess the proof is in the pudding, or in golf it's the putting.

I have only a rationalization: in `package p`, I know nothing is inherited in this scope (because we do not `package object p extends T`). Once someone defines a class, the folks working in that mine must manage that namespace, where stuff is inherited.

I agree there is regularity in seeing `object p` as just a way to "associate" some definitions, and I think they speak of "package membership", but if the regularity were borne out, then we would love `package object p extends T`. You would be asking, Why is that different from `object p extends T`?

But all that is just a rationalization.

What are some examples of pernicious shadowing of top-level defs? I can't wait to see.

Comment (lrytz) [reply spec/02-identifiers-names-and-scopes.md:20]: An example was:

```scala
package p

class Code extends Parent {
  def impl = s"The $CommonName"
}

case class CommonName(some: Thing)
```

A second file:

```scala
package p
trait Parent {
  val CommonName = "something"
}
```

It's maybe clear here that we're not intending to refer to the synthetic case class companion, but this could still be a head scratcher for someone looking at the first file. One could for example do a search in the editor for `CommonName` and land on the case class.

In this case I moved `CommonName` to a separate file instead of changing the reference.

Comment (som-snytt) [reply spec/02-identifiers-names-and-scopes.md:20]: Aha, that does seem lintable, at least with `CommonName` in the same file. Conversely, it loses precedence when moved to a different file, but the reference in `Parent` does not.

I just realized this feature is not released until 2.13.11. Let me take a moment to clear my fever-addled brain. 

I notice `toString` is not exempt.

Also I was about to wonder if the use case is really anon classes, esp SAM types. The user expectation is more obviously, "I am focused on implementing that one method, and probably unaware of inheriting anything."

Comment (lrytz) [reply spec/02-identifiers-names-and-scopes.md:20]: > `CommonName` in the same file

Yes, I mean to warn only if the outer definition is in the same compilation unit.

> I notice `toString` is not exempt.
> if the use case is really anon classes

Those are very good points. I certainly saw it pop up a lot with anonymous classes, and one would expect `toString` to resolve in the current object. Hmm. Since this is a spec change and not an impl detail (a lint rule), we need to have general and hopefully simple rules. An exception for `toString` wouldn't work...

Comment (lrytz) [reply spec/02-identifiers-names-and-scopes.md:20]: I think we can wait to see if `toString` really is problematic. So far (Scala 3, my tests of the new Scala 2 warning) it didn't seem so.

**Comment by SethTisue** [comment]
the "under `-Xsource:3`" part happened in the followup PR, #10339

**Comment by SethTisue** [comment]
@lrytz actually maybe I'm confused and maybe the edit to the issue title I just did is wrong? if so, can you change it back?

and what is the appropriate text to include in the 2.13.11 release notes to cover both this PR and #10339? can you suggest an appropriate edit over at https://github.com/scala/scala-dev/pull/843 ?

**Comment by som-snytt** [comment]
@SethTisue I took a shot and trivially proposed using the title, which covers both PRs.
